src/ephemeris.rs

```diff
@@ -45,18 +46,32 @@ pub enum Status {
     TooOld,
     InvalidSid,
     InvalidIod,
+}
+
+/// Various statuses that an ephemeris can be in
+#[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub enum Status {
+    Invalid(InvalidEphemeris),
```

Comment:
This feels like it could be a modeled as Result<(), InvalidEphemeris>. Reason I mention it, is there are some if lets where you check for invalid status’s and propagate them. That could probably be done with ? if this error checking was Result based. A .to_result() method (like .ok() on Option) would also probably work.

Reply:
I guess that’s exactly what EphemerisError is, so maybe my question is better phrased as why do we need this separate status enum, when Result<(),EphemerisError> should be functionally the same?

Reply:
This is one the primary places where I duplicate a function call that the C code is doing and also was where I'm less sure about my choices, I'm glad you pointed this out.

When evaluating the ephemeris the C code checks the status of the ephemeris and if it's not valid it returns a generic error value. In the Rust interface I thought it would be nicer to return a specific reason why the evaluation failed. Duplicating the exact status enums from the C code means having both `Valid` and all invalid statuses listed. When calling `get_status()` this seems ideal, you get a single enum value one of which is valid and all others are invalid. When evaluating the ephemeris potentially returning `Valid` as an error state seems not great. To accommodate both use cases I split the enum into two, as it is here, one listing just the invalid variants and another encapsulating the invalid variants along with the valid variant.

Returning `Result` from `get_status()` doesn't seem "right" because `Result` implies potential for a failed function call. If the ephemeris isn't valid the call to `get_status()` itself hasn't failed, it's only meant to report on the state of the ephemeris. This is doubly true when you consider that ephemeris validity is partly based on the point in time you evaluate it at (i.e. an ephemeris could be valid at noon but invalid an hour later). Maybe some helper functions like `.to_result` and getting rid of `EphemerisError` and just impl Error for `InvalidEphemeris` directly is preferable?

Reply:
> Returning Result from get_status() doesn't seem "right" because Result implies potential for a failed function call. If the ephemeris isn't valid the call to get_status() itself hasn't failed, it's only meant to report on the state of the ephemeris. This is doubly true when you consider that ephemeris validity is partly based on the point in time you evaluate it at (i.e. an ephemeris could be valid at noon but invalid an hour later).

That makes a lot of sense! Not really sure what the best move is. I think a `to_result` might be nice. The only reason I mention it is there are a few places where this happens - 

```rust
let status = self.get_detailed_status(t);
if let Status::Invalid(invalid_status) = status {
    return Err(EphemerisError(invalid_status));
}
```

with a `to_result` function on `Status` we could get something like - 

```rust
let status = self.get_detailed_status(t).to_result()?;
or
let status = self.get_detailed_status(t).ok()?;
```

Which seems nice. Or maybe there could be two `get_status` functions. One that behaves as is, and one that returns an error if the ephemeris isn't valid. Like a `try_check_status` and `check_status`. Not really sure, think I prefer the first approach - a way to convert to a Result.

Reply:
I've added a `to_result` function and made the changes you suggested. I think I'll leave things like that for now. If/when we use this we can see if there's further improvements we can make.